flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SynchronousCheckpointITCase.java

```diff
@@ -125,16 +126,24 @@ public void taskCachedThreadPoolAllowsForSynchronousCheckpoints() throws Excepti
 			assertEquals(ExecutionState.RUNNING, task.getExecutionState());
 			assertEquals(CheckpointingState.NONE, synchronousCheckpointPhase.getState());
 
+			// Hack: we are triggering a checkpoint with advanceToEndOfEventTime = true, to be sure that
+			// triggerCheckpointBarrier has reached the sync checkpoint latch (by verifying in
+			// SynchronousCheckpointTestingTask.advanceToEndOfEventTime) and only then proceeding to
+			// notifyCheckpointComplete.
+			// Without such synchronization, the notifyCheckpointComplete execution may be executed first and leave this
+			// test in a deadlock.
 			task.triggerCheckpointBarrier(
 					42,
 					156865867234L,
 					new CheckpointOptions(CheckpointType.SYNC_SAVEPOINT, CheckpointStorageLocationReference.getDefault()),
-					false);
+					true);
```

Comment:
Instead of that, could we return a `CompletableFuture` from the stack trace:
```
Task#triggerCheckpointBarrier ->
Task#executeAsyncCallRunnable -> {
	(...)
	try {
		return executor.submit(runnable, blocking);
	}
	(...)
```
and wait for the future to complete before calling `notifyCheckpointComplete`?

Reply:
Do you mean to get a future that would wait for result of `invokable.triggerCheckpoint()` call?
In that case, I think what you are proposing would deadlock the test.

Reply:
Yes, you are right. I forgot that the code is waiting on the latch there.